pkg/ddc/juicefs/sync_runtime.go

```diff
@@ -70,33 +80,108 @@ func (j *JuiceFSEngine) syncWorkerSpec(ctx cruntime.ReconcileRequestContext, run
+			workersToUpdate.Spec.Template.Spec.Volumes = newVolumes
+			changed = true
+		}
+
+		// labels
+		if labelChanged, newLabels := j.isLabelsChanged(workersToUpdate.Spec.Template.ObjectMeta.Labels, value.Worker.Labels); labelChanged {
+			workersToUpdate.Spec.Template.ObjectMeta.Labels = newLabels
+			changed = true
+		}
+
+		// annotations
+		if annoChanged, newAnnos := j.isAnnotationsChanged(workersToUpdate.Spec.Template.ObjectMeta.Annotations, value.Worker.Annotations); annoChanged {
+			workersToUpdate.Spec.Template.ObjectMeta.Annotations = newAnnos
+			changed = true
+		}
+
+		// options -> configmap
+		workerCommand, err := j.getWorkerCommand()
+		if err != nil || workerCommand == "" {
+			j.Log.Error(err, "Failed to get worker command")
+			return err
+		}
+		cmdChanged, _ = j.isCommandChanged(workerCommand, value.Worker.Command)
+
 		if len(workersToUpdate.Spec.Template.Spec.Containers) == 1 {
-			workerResources := runtime.Spec.Worker.Resources
-			if !utils.ResourceRequirementsEqual(workersToUpdate.Spec.Template.Spec.Containers[0].Resources, workerResources) {
-				j.Log.Info("The resource requirement is different.", "worker sts", workersToUpdate.Spec.Template.Spec.Containers[0].Resources, "runtime", workerResources)
-				workersToUpdate.Spec.Template.Spec.Containers[0].Resources = workerResources
+			// resource
+			if resourcesChanged, newResources := j.isResourcesChanged(workersToUpdate.Spec.Template.Spec.Containers[0].Resources, runtime.Spec.Worker.Resources); resourcesChanged {
+				workersToUpdate.Spec.Template.Spec.Containers[0].Resources = newResources
 				changed = true
-			} else {
-				j.Log.V(1).Info("The resource requirement of workers is the same, skip")
 			}
 
-			if changed {
-				if reflect.DeepEqual(workers, workersToUpdate) {
-					changed = false
-					j.Log.V(1).Info("The resource requirement of worker is not changed, skip")
-					return nil
-				}
-				j.Log.Info("The resource requirement of worker is updated")
+			// env
+			if envChanged, newEnvs := j.isEnvsChanged(workersToUpdate.Spec.Template.Spec.Containers[0].Env, value.Worker.Envs); envChanged {
+				workersToUpdate.Spec.Template.Spec.Containers[0].Env = newEnvs
+				changed = true
+			}
+
+			// volumeMounts
+			if volumeMountChanged, newVolumeMounts := j.isVolumeMountsChanged(workersToUpdate.Spec.Template.Spec.Containers[0].VolumeMounts, value.Worker.VolumeMounts); volumeMountChanged {
+				workersToUpdate.Spec.Template.Spec.Containers[0].VolumeMounts = newVolumeMounts
+				changed = true
+			}
+
+			// image
+			runtimeImage := value.Image
+			if value.ImageTag != "" {
+				runtimeImage = runtimeImage + ":" + value.ImageTag
+			}
+			if imageChanged, newImage := j.isImageChanged(workersToUpdate.Spec.Template.Spec.Containers[0].Image, runtimeImage); imageChanged {
+				workersToUpdate.Spec.Template.Spec.Containers[0].Image = newImage
+				changed = true
+			}
+		}
 
+		if cmdChanged {
```

Comment:
If both `cmdChanged` and `changed`, will it update the workersToUpdate twice? I think it may fail because the runtime cache will not update so soon?

Reply:
If both cmdChanged and changed, sts will not rollout restart. That is to say, will not update twice.